### PR TITLE
feat: implement catalog syncing for `publish`

### PR DESCRIPTION
**Pull request by ysndr**
This PR implements logic to publish catalogs to branches on remote repositories.

Since flox and thus publishing is `git` based, we have to

1. clone a repo
2. checkout or create the relevant catalog branch (setting up the push remote)
3. ensure the snapshot does not exist yet (?)
4. write the snapshot to the repo
5. commit the snapshot file
6. push the catalog branch to the upstream repo

This PR abstracts these steps over our Git abstraction.

`UpstreamRepo` is a thin layer allowing us to clone a repo and create `UpstreamCatalog` instances that represent an individual catalog branch.
`UpstreamCatalog` then implements operations on catalogs i.e. retrieving, writing and committing snapshots, as well as pushing the snapshot back to origin.

The two types allow safe interactions with a catalog.
It is only possible to have a single catalog branch checked out through `UpstreamCatalog` since a `mut` reference of `UpstreamRepo` is required to construct a catalog instance.
`UpstreamCatalog` enforces synching to upstream before a new branch can be checked out, to avoid leaving the repository in a bad state.
`UpstreamRepo` is only constructed from a fresh clone, so no two instances conflict.

### CODE REVIEW

**Comment by ysndr** [comment]
based on https://github.com/flox/flox/issues/140
Ping @zmitchell, @mkenigs

**Review by mkenigs — CHANGES_REQUESTED** [review]
I don't quite understand why we're doing this. Weren't we going to leave pushing to the user?

**Comment by ysndr** [comment]
The manual push was mentioned under the premise that we do "local only" publishes with the catalog on the main branch where users have a greater interest to control the commit history.

When publishing to catalog branches we should avoid doing that in the same repository, while possible its really awkward committing to commit to branches without a checkout, and checking out unrelated branches means you change around the state of the current repo which can have all kinds of git failures.

Its also easier to base the publishing fully on remote repos because there we are guaranteed the repo is accessible, the published commit is upstream and after a publish the package is available.

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 1]
```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
...
+            Err(e) if e.kind() == io::ErrorKind::NotFound => None,
+            Err(e) => Err(e)?,
+        };
+
+        Ok(read_snapshot)
+    }
+
+    /// [Value] to be [flox_types::catalog::CatalogEntry]
+    ///
+    /// Consumers should check if a snapshot already exists with [Self::get_snapshot]
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:324]: Why not perform the check here? Also no caller is using the return value of `get_snapshot`, so it could be a `contains_snapshot() -> bool` method instead

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 2]
```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
...
+            .join(snapshot["eval"]["meta"]["pname"].to_string());
+        snapshot["eval"]["meta"]["version"].to_string();
+
+        path.set_file_name(format!("{}.json", snapshot["eval"]["meta"]["version"]));
+        path
+    }
+
+    /// Try retrieving a snapshot from the catalog
+    /// TODO: better addressing (attrpath + version + drv hash?)
+    fn get_snapshot(&self, snapshot: &Value) -> Result<Option<Value>, PublishError> {
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:311]: I'm guessing you're hoping to replace `Value` with `CatalogEntry` throughout?

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 3]
```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
...
+    fn catalog_branch_name(system: &System) -> String {
+        format!("catalog/{system}")
+    }
+
+    /// Create an [UpstreamCatalog] by checking out or creating a catalog branch.
+    ///
+    /// `Git` objects can switch branches at any time leaving the repo in an unknown state.
+    /// [get_catalog] ensures that only one [UpstreamCatalog] exists at a time by requiring a `&mut self`.
+    async fn get_catalog(&mut self, system: &System) -> Result<UpstreamCatalog, PublishError> {
+        if self.0.list_branches().await? // todo: catch error
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:268]: Listing all branches is not the most efficient approach but not something we need to block on

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 4]
```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
...
+    }
+
+    fn catalog_branch_name(system: &System) -> String {
+        format!("catalog/{system}")
+    }
+
+    /// Create an [UpstreamCatalog] by checking out or creating a catalog branch.
+    ///
+    /// `Git` objects can switch branches at any time leaving the repo in an unknown state.
+    /// [get_catalog] ensures that only one [UpstreamCatalog] exists at a time by requiring a `&mut self`.
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:266]: Does that guarantee that a method in UpstreamCatalog doesn't call checkout?

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 2 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:311]: The result, yes
the address (`snapshot: &Value`) I'd like to more lightweight

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 5]
```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
...
+        }
+        Ok(UpstreamCatalog(&self.0))
+    }
+}
+
+/// Representation of a specific catalog branch in an exclusive clone of an upstream repo.
+///
+/// [UpstreamCatalog] guaranteesd that during its lifetime all operations on the underlying git repo
+/// are performed on a single branch,
+/// and that the branch is pushed to upstream before a new branch can be checked out.
```
[the remark this answers is not in the document]
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:290]: Why do you think its undesirable?
I can only think of publishing multiple systems at once as a use case to checkout other branches but flox is squarely running with a single system context, and mutliple systesm was not a requirement of publish so far.


If we want this it should be easy to implement it by running an `--atomic` push of all (or all modified) catalog branches on the `UpstreamRepo` type and return back control with `add_snapshot` or a separate method that destructs `UpstreamCatalog`

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 1 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:324]: I didn't want to restrict behaviour around republishing/overwriting at this level that should be a decision of the Application level.

`get_snapshot` returns a snapshot so we can do updates on them rather than blindly overriding. -- I do concede, none of that is implemented and our application check would likely be a `get_snapshot()?.is_some()`.

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 4 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:266]: yes, `&mut` references are enforced to be unique by the compiler.

Because `UpstreamCatalog` holds a reference from a mutable reference on `UpstreamCatalog`, `UpstreamCatalog` can't be borrowed until the catalog with that reference is dropped.

The alternative is to use move semantics and move `UpstreamRepo` into the type, i.e. 

```diff 
- struct UpstreamCatalog<'a>(&'a Git);
+ struct UpstreamCatalog<'a>(&'a Git, UpstreamRepo);

- fn get_catalog(&mut self, system: &System) -> Result<UpstreamCatalog, PublishError>
+ fn get_catalog(self, system: &System) -> Result<UpstreamCatalog, PublishError>

```

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 6]
```diff
@@ -215,27 +217,17 @@ impl<'flox> Publish<'flox, NixAnalysis> {
...
-            .into_iter().any(|info| info.name == "catalog")
-        {
-            catalog.checkout("catalog", false).await.unwrap(); // todo: catch error
-        } else {
-            todo!();
-            // catalog.checkout("catalog", true).await.unwrap(); // todo: catch error
-
-            // catalog.set_upstream("origin", "catalog").await.unwrap();  // todo: implement
-            //                                                               todo: catch error
+    pub async fn push_snapshot(&self) -> Result<(), PublishError> {
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/publish.rs:226]: Hmm yeah we probably need a better name. Snapshot doesn't really say what it is either - it could be a snapshot of anything. No need to bikeshed here though

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/publish.rs:296]: Depending on how we setup builders, I would expect a possibly quite common use case for publish would be to run `flox publish --systems x86_64-linux aarch64-darwin` in CI. I agree that's not a requirement at the moment, but it seems like creating abstractions to explicitly prevent something we'll likely want to do in the future is a waste of effort. That said maybe it will make things safer for now. Doesn't have to be a blocker.

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 2 above)
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/publish.rs:317]: But you'd expect to replace `snapshot["eval"]["meta"]["pname"].to_string()` above at some point?

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:296]: Doing multiple system publishes would change the flow anyway.
Changing when to drop `CatalogChannel` is a trivial change compared to that.

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 2 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:317]: Yes the current `get_snapshot_path` is very preliminary and subject to change based on what we decide for the catalog to be like.

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 4 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:272]: Sure if the implementation of `UpstreamCatalog` performs a checkout, that's nothing that is prevented by this.
The point of this guarantee is that you can't get two `UpstreamCatalog` instances of the same repo at the same time from an `UpstreamRepo`

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 7]
```diff
@@ -223,42 +237,171 @@ impl<'flox> Publish<'flox, NixAnalysis> {
...
+        temp_dir: impl AsRef<Path>,
+    ) -> Result<Self, PublishError> {
+        let repo_dir = tempfile::tempdir_in(temp_dir).unwrap().into_path(); // todo catch error
+        let repo = <Git as GitProvider>::clone(url.as_ref(), &repo_dir, false).await?;
+
+        Ok(Self(repo))
+    }
+
+    fn catalog_branch_name(system: &System) -> String {
+        format!("catalog/{system}")
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:283]: I was wondering today, will we support catalogs for different branches? E.g. if I have two branches nixpkgs-stable and nixpkgs-unstable in the same repo, should we create `nixpkgs-stable/catalog/{system}` and `nixpkgs-unstable/catalog/{system}` so there isn't a collision? No need to change that now, just thought I'd mention the question

**Review by mkenigs — COMMENTED** [review]
Clearing request for changes

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 8]
```diff
@@ -754,4 +897,168 @@ mod tests {
...
+        .expect("Should clone repo");
+
+        assert!(repo.0.list_branches().await.unwrap().is_empty());
+
+        let catalog = repo
+            .get_or_create_catalog(&"aarch64-darwin".to_string())
+            .await
+            .expect("Should create branch");
+
+        // commit a file to the branch to crate the first reference
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:948]: Why is that necessary?

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 9]
```diff
@@ -754,4 +897,168 @@ mod tests {
...
+            .await
+            .expect("Should create branch");
+
+        catalog
+            .add_snapshot(&snapshot)
+            .await
+            .expect("Should add snapshot");
+        catalog.push_catalog().await.expect("Should push catalog");
+
+        // checkout the new brnach upstream to check if the file got written
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:1047]:
```suggestion
        // checkout the new branch upstream to check if the file got written
```

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 10]
```diff
@@ -754,4 +897,168 @@ mod tests {
...
+            repo.workdir().unwrap().to_string_lossy(),
+            temp_dir_handle.path(),
+        )
+        .await
+        .expect("Should clone repo");
+
+        assert!(repo.0.list_branches().await.unwrap().is_empty());
+
+        let catalog = repo
+            .get_or_create_catalog(&"aarch64-darwin".to_string())
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:944]: Would it be better to use `flox.system` instead of hardcoding? Probably doesn't really matter

File: crates/flox-rust-sdk/src/models/publish.rs [hunk 11]
```diff
@@ -223,42 +237,171 @@ impl<'flox> Publish<'flox, NixAnalysis> {
...
+/// are performed on a single branch,
+/// and that the branch is pushed to upstream before a new branch can be checked out.
+struct UpstreamCatalog<'a>(&'a Git);
+
+impl UpstreamCatalog<'_> {
+    /// Mostly naïve approxiaton of a snapshot path
+    ///
+    ///  /packages/<pname>/<version>.json
+    ///
+    /// TODO: fix before releasing publish!
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/publish.rs:321]: Fix what?
```suggestion
    /// TODO: fix before releasing publish!
```

**Review by mkenigs — APPROVED** [review]

File: crates/flox-rust-sdk/src/models/publish.rs (hunk 8 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:948]: we're using orphan branches. unless you create a reference a branch can point to, the branch does not actually exist yet, with normal branches they inherit the current HEAD ref

Comment (ysndr) [reply crates/flox-rust-sdk/src/models/publish.rs:948]:
```suggestion
        // commit a file to the branch to crate the first reference on the orphan branch
```
